File: Subtree of Another Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
                if (root == None) :
            return False
                # print(root.val)
        return self.isSubtree(root.left,subRoot) or (root.val == subRoot.val and self.check(root,subRoot)) or self
            .isSubtree(root.right,subRoot)
            def check(self,root,subroot):
        # print(root.val, subroot.val)
        if root==None and subroot == None:
            return True
        elif (root and not subroot) or (subroot and not root):
            return False
                return root.val == subroot.val and self.check(root.left,subroot.left) and self.check(root.right,subroot
            .right)

File: Subtree of Another Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
        bool isSubtree(TreeNode* s, TreeNode* t) {
       return isSubtree(s,t,false);
    }  
         bool isSubtree(TreeNode* s, TreeNode* t, bool f) {
        if (!s && !t ) return true;
        else if ((s&&!t) || (t&&!s) ) return false;
        return ((s->val == t->val) && isSubtree(s->left, t->left,true)&&isSubtree(s->right, t->right,true))
                || (!f && (isSubtree(s->left, t,false) || isSubtree(s->right, t,false)) );
    }
};